src/target/xtensa/xtensa.c

```diff
@@ -1392,18 +1530,11 @@ int xtensa_do_step(struct target *target, int current, target_addr_t address, in
 				"disabling IRQs while stepping is not implemented w/o high prio IRQs option!");
 			return ERROR_FAIL;
 		}
-		/* Mask all interrupts below Debug, i.e. PS.INTLEVEL = DEBUGLEVEL - 1 */
-		xtensa_reg_val_t temp_ps = (oldps & ~0xF) | (xtensa->core_config->debug.irq_level - 1);
-		xtensa_reg_set(target, XT_REG_IDX_PS, temp_ps);
+		/* Update ICOUNTLEVEL accordingly */
+		icountlvl = MIN((oldps & 0xF) + 1, xtensa->core_config->debug.irq_level);
```

Comment:
As far as I understand the idea of `maskisr on` (what corresponds to `XT_STEPPING_ISR_OFF`) is to disable IRQ processing at all. At least it looks it is implemented in this way for ARM. This change allows IRQ to happen and just hides them for debugger.
I think we need to add special mode `auto` for stepping with enabled interrupts, but do not be disturbed by them as [it is done for Cortex-M](https://github.com/espressif/openocd-esp32/blob/master/doc/openocd.texi#L10244).

Reply:
I think the confusion here is that the original Espressif implementation used to mask ISRs entirely (as an option) during step, by changing ps.intlevel.  Since that behavior can be somewhat destructive to application flow, I reimplemented it to use icountlevel properly, so that the interrupt always still occurs--now it is either taken behind the scenes (and many instructions are run during the step) or it is taken and stepping lands at the first vector instruction.  So we're no longer masking the interrupt entirely, but rather the GDB user's ability to land in it.  But sure, if there is value in keeping all 3 options, we can add one back in the future.

Reply:
Ok. In future I think we can implement all modes defined for [Cortex-M}(https://github.com/espressif/openocd-esp32/blob/master/doc/openocd.texi#L10244):
- `on` - IRS are always enabled during stepping and resume
- `off` - IRS are always disabled during stepping and resume
- `steponly` - IRS are always disabled during stepping only. That corresponds to Espressif original implementation
- `auto` - IRS are always enabled during stepping  and resume, but debugger does not step into ISRs. That corresponds to implementation in this PR.

Reply:
@erhankur This may need updating one of our stepping test

Reply:
Sure, matching our interrupt stepping options with Coretex-M makes sense.  Please let me know if you need this sooner for any reason, otherwise it's on my list for upcoming features.